### PR TITLE
Bug fix: Use target tablet from health stats cache when checking replication status

### PR DESCRIPTION
**Pull request by austenLacy**
## Description

When determining which tablets to show in `show vitess_replication_status` Vitess is excluding tablets if their states were changed. This now uses the target tablet from the heath check cache which represents the expected state of the tablets. 

## Related Issue(s)

- Fixes: https://github.com/vitessio/vitess/issues/14432

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/test/endtoend/cluster/cluster_process.go [hunk 1]
```diff
@@ -985,6 +985,15 @@ func (cluster *LocalProcessCluster) VtctlclientGetTablet(tablet *Vttablet) (*top
 	return &ti, nil
 }
 
+func (cluster *LocalProcessCluster) VtctlclientChangeTabletType(tablet *Vttablet, tabletType topodatapb.TabletType) error {
+	_, err := cluster.VtctlclientProcess.ExecuteCommandWithOutput("ChangeTabletType", "--", tablet.Alias, tabletType.String())
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (mattlord) [inline go/test/endtoend/cluster/cluster_process.go:994]: Can just return err here.

File: go/test/endtoend/tabletgateway/vtgate_test.go [hunk 2]
```diff
@@ -69,6 +70,34 @@ func TestVtgateReplicationStatusCheck(t *testing.T) {
 	assert.Equal(t, expectNumRows, numRows, fmt.Sprintf("wrong number of results from show vitess_replication_status. Expected %d, got %d", expectNumRows, numRows))
 }
 
+func TestVtgateReplicationStatusCheckWithTabletTypeChange(t *testing.T) {
+	defer cluster.PanicHandler(t)
+	// Healthcheck interval on tablet is set to 1s, so sleep for 2s
+	time.Sleep(2 * time.Second)
```
Comment (mattlord) [inline go/test/endtoend/tabletgateway/vtgate_test.go:76]: This is likely to be flaky in the CI due to unpredictable performance and occasionally machine pauses. Any reason not to use the variable value (if we can access it) * 10 or something?

**Review by mattlord — COMMENTED** [review]
Looks good. Thanks, @austenLacy ! ❤️ 

I only had a couple of minor comments -- let me know what you think. Have you already run the new test locally in a loop to check for flakiness?

I can come back to this quickly when I hear back.

File: go/vt/vtgate/executor.go [hunk 3]
```diff
@@ -901,7 +901,7 @@ func (e *Executor) showVitessReplicationStatus(ctx context.Context, filter *sqlp
 	for _, s := range status {
 		for _, ts := range s.TabletsStats {
 			// We only want to show REPLICA and RDONLY tablets
-			if ts.Tablet.Type != topodatapb.TabletType_REPLICA && ts.Tablet.Type != topodatapb.TabletType_RDONLY {
+			if ts.Target.TabletType != topodatapb.TabletType_REPLICA && ts.Target.TabletType != topodatapb.TabletType_RDONLY {
```
Comment (mattlord) [inline go/vt/vtgate/executor.go:904]: I did confirm that `Target.TabletType` is what's being used elsewhere when examining/showing the type for tablets in the healthcheck cache. So this part seems right.

Following on from that, I think that we should change all usage of `ts.Tablet.*` in this function to `ts.Target.*`. For example, below we should change `ts.Tablet.Keyspace` to `ts.Target.Keyspace` and `ts.Tablet.Shard` to `ts.Target.Shard`.

**Review by mattlord — COMMENTED** [review]
Found one more issue that I think we should address after having looked at other usage of `TabletsStats`.

@deepthi could you please review this when you have time? No rush, but you know the healthcheck cache related code the best. Thanks!

File: go/test/endtoend/tabletgateway/vtgate_test.go (hunk 2 above)
Comment (austenLacy) [reply go/test/endtoend/tabletgateway/vtgate_test.go:76]: I copied this from the other tests in this file. I can try running the test in a loop to see if it has any flakiness.

Comment (austenLacy) [reply go/test/endtoend/tabletgateway/vtgate_test.go:76]: I ran it 10 times in a loop 2-3 times and didn't see any flakiness.

```
for i in {1..10}; do go test -count=1 -timeout 30s -run ^TestVtgateReplicationStatusCheckWithTabletTypeChange$ vitess.io/vitess/go/test/endtoend/tabletgateway; sleep 1; done
ok      vitess.io/vitess/go/test/endtoend/tabletgateway 23.902s
ok      vitess.io/vitess/go/test/endtoend/tabletgateway 25.247s
ok      vitess.io/vitess/go/test/endtoend/tabletgateway 23.130s
ok      vitess.io/vitess/go/test/endtoend/tabletgateway 23.091s
ok      vitess.io/vitess/go/test/endtoend/tabletgateway 25.394s
ok      vitess.io/vitess/go/test/endtoend/tabletgateway 25.882s
ok      vitess.io/vitess/go/test/endtoend/tabletgateway 23.308s
ok      vitess.io/vitess/go/test/endtoend/tabletgateway 23.246s
ok      vitess.io/vitess/go/test/endtoend/tabletgateway 25.528s
ok      vitess.io/vitess/go/test/endtoend/tabletgateway 25.797s
```

**Review by mattlord — APPROVED** [review]
Thanks, @austenLacy ! ❤️

**Review by deepthi — APPROVED** [review]
Nice work on tests!

File: go/vt/vtgate/executor.go [hunk 4]
```diff
@@ -901,14 +901,14 @@ func (e *Executor) showVitessReplicationStatus(ctx context.Context, filter *sqlp
...
```
Comment (deepthi) [inline go/vt/vtgate/executor.go:904]: It's actually worrisome that these two fields are not in sync. That is something we'll need to go back and review. Doesn't need to block this PR though.

**Pull request merged**